pkg/services/authn/authnimpl/sync/oauth_token_sync.go

```diff
@@ -65,8 +70,12 @@ func (s *OAuthTokenSync) SyncOauthTokenHook(ctx context.Context, identity *authn
 	}
 
 	expires := token.OAuthExpiry.Round(0).Add(-oauthtoken.ExpiryDelta)
+	idTokenHasExpired, err := hasIDTokenExpired(token)
+	if err != nil {
+		s.log.FromContext(ctx).Error("Failed to check expiry of ID token", "id", identity.ID, "error", err)
+	}
 	// token has not expired, so we don't have to refresh it
-	if !expires.Before(time.Now()) {
+	if !expires.Before(time.Now()) && !idTokenHasExpired {
 		// cache the token check, so we don't perform it on every request
 		s.cache.Set(identity.ID, struct{}{}, getOAuthTokenCacheTTL(expires))
```

Comment:
To consider both tokens' expiry and to refresh those when needed the [getOAuthTokenCacheTTL](https://github.com/grafana/grafana/pull/69829/files#diff-ecd19dccd5583801dfd8752e347c08427fa81deba3285c6e7e17ccba016cd382R105) function should also consider the expiry of the `id_token` and set the TTL as the `min(AT expiry, IT expiry)`. If IT expiry `IsZero` then only use the AT expiry.

An edge case could be when the AT expiry is Zero (never expiring) and an id_token is provided (`exp` is mandatory for id_tokens), but I don't think it could happen, at least not for the supported OAuth integrations.